backend/store/memory/event_store.go

```diff
@@ -0,0 +1,366 @@
+	return &memorydb{
+		limiter: limiter,
+	}
+}
+
+type eventEntry struct {
+	// Serialized and compressed event
+	EventBytes []byte
+
+	// Dirty indicates if the entry has yet to be written out to long term storage
+	Dirty bool
+
+	// Mu is necessary in the rare instance when multiple goroutines are competing
+	// to store different versions of the same event. This is only necessary due to
+	// a design flaw in eventd, and could possibly be mitigated there.
+	Mu sync.Mutex
+}
+
+const eventOutputTruncatedBytesLabel = "sensu.io/output_truncated_bytes"
+
+type SilenceStoreI interface {
+	// DeleteSilences deletes all silences matching the given names.
+	DeleteSilences(ctx context.Context, namespace string, names []string) error
+
+	// GetSilences returns all silences in the namespace. A nil slice with no error is
+	// returned if none were found.
+	GetSilences(ctx context.Context, namespace string) ([]*corev2.Silenced, error)
+
+	// GetSilencedsByCheck returns all silences for the given check . A nil
+	// slice with no error is returned if none were found.
+	GetSilencesByCheck(ctx context.Context, namespace, check string) ([]*corev2.Silenced, error)
+
+	// GetSilencedEntriesBySubscription returns all entries for the given
+	// subscription. A nil slice with no error is returned if none were found.
+	GetSilencesBySubscription(ctx context.Context, namespace string, subscriptions []string) ([]*corev2.Silenced, error)
+
+	// GetSilenceByName returns an entry using the given namespace and name. An
+	// error is returned if the entry is not found.
+	GetSilenceByName(ctx context.Context, namespace, name string) (*corev2.Silenced, error)
+
+	// UpdateSilence creates or updates a given silence.
+	UpdateSilence(ctx context.Context, silence *corev2.Silenced) error
+
+	// GetSilencesByName gets all the named silence entries.
+	GetSilencesByName(ctx context.Context, namespace string, names []string) ([]*corev2.Silenced, error)
+}
+
+func (m *memorydb) ReadEntry(namespace, entity, check string) *eventEntry {
+	key := strings.Join([]string{namespace, entity, check}, "\n")
+	emptyEntry := &eventEntry{}
+	result, _ := m.data.LoadOrStore(key, emptyEntry)
+	entry := result.(*eventEntry)
+	return entry
+}
+
+func (m *memorydb) WriteTo(ctx context.Context, s store.EventStore) error {
+	var storeErr error
+	ctx = store.NoMergeEventContext(ctx)
+	m.data.Range(func(key, value any) bool {
+		entry := value.(*eventEntry)
+		entry.Mu.Lock()
+		defer entry.Mu.Unlock()
+		if !entry.Dirty {
+			return true
+		}
+		eventBytes, err := snappy.Decode(nil, entry.EventBytes)
+		if err != nil {
+			// developer error, fatal
+			panic(err)
+		}
+		var event corev2.Event
+		if err := proto.Unmarshal(eventBytes, &event); err != nil {
+			// developer error, fatal
+			panic(err)
+		}
+		if err := m.limiter.Wait(ctx); err != nil {
+			// context cancellation
+			return false
+		}
+		_, _, storeErr = s.UpdateEvent(ctx, &event)
+		if storeErr != nil {
+			return false
```

Comment:
I'm wondering if it's not worth having a way to still try to store the rest of the events even if one of them fails to be stored as we range over the map.

I know it makes tracking and returning errors more complex, but my main concern is what happens if we stop going over the map early as we do here: a single problematic event blocks all the other events "behind it in the map" from being persisted.

Would it not be better if that single problematic event didn't have such an impact on the other values to be persisted? At least the system would still mostly work rather than having to somehow do something with the problematic event so that the rest of the system keeps working.

Reply:
At this point the event is validated, so it's not likely to be a problematic individual event, but rather a systemic problem with the event store, like perhaps the connection has failed or something. In that case it's best not to keep trying and wait for the next write-out window.

Reply:
That's fair, thank you.